src/prefect/task_engine.py

```diff
@@ -640,6 +638,19 @@ def start(
     def transaction_context(self) -> Generator[Transaction, None, None]:
         result_factory = getattr(TaskRunContext.get(), "result_factory", None)
 
+        def __prefect_task_rollback_handler(txn: Transaction):
+            # if there are no user-defined rollback hooks to run, then don't
+            # explicitly mark as rolled back
+            if txn.on_rollback_hooks == [__prefect_task_rollback_handler]:
+                return
```

Comment:
I added this after noticing that since tasks without any user-supplied rollback hooks still technically are rolled back when an overarching transaction is rolledback, this would update all tasks to have a RolledBack state. 

I'm thinking that from a user's perspective, if the task isn't registered to _do anything_ on rollback, then calling it rolled back could be a bit more misleading than if it were just in its traditional completed state. 

For example, consider a task that has retries. On failure, since each task is its own transaction, it's actually rolled back before being retried so without any rollback hooks, the task would enter Completed(name="RolledBack"). 

Soft opinion to follow prior behavior here when a user hasn't opted into transactional rollback semantics.

Reply:
I get what you mean, but I think this is false precision for the feature.

Just to illustrate a confusion that could arise with this implementation: imagine three tasks in a transaction, and task one and task three both have rollback hooks.  In the UI, you'll see the first and last tasks were `"RolledBack"` but the second one wasn't? or did something else happen? etc.

Reply:
👀 
On removal of this short circuit, I started working through some of the test failures I was originally trying to avoid and I think I hear ya on the _imprecision_ here. What I had before misses the world where a task runs to completion and is then rolled back as part of an overall transaction (your example), but removing this short circuit entirely misses a task failing "normally" ends up in Running -> RolledBack -> Failed which is _technically_ correct but misleading since that task's user defined rollback hooks don't run in that scenario since they never got staged on success. 

I'm now thinking this should be:
```python
if txn.state != TransactionState.STAGED:
    return
...
```

🧐